text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,749 @@
+* This minimal version would considerably reduce the burden on the network of new nodes joining, at the cost of periodic activity spikes on nodes pushing snapshots to a CDN.
+* Another approach is to allow any node to push objects to a CDN. This might require a plausibility check (i.e. show that the object is reachable from genesis) and thus would need some specialized software to enable.
+* Clients could also upload Wasm blocks to a CDN beforehand, significantly reducing the size of a referencing Deploy.
+
+_NOTE: the described CDN functionality would be strictly optional with regards to the correct operation of the network._
+
+### Git-style pack files
+
+We could opt to implement pack files, to further optimize and most notably reduce total round trips. This is a non-trivial optimization, but offers significant benefits. For instance, it is easy to imagine a new node receiving a pack file that contains all Blocks (without dependencies) all the way back to genesis; that node could then fetch other information as required.
+
+A pack is just a collection of objects, so importing one into a node should be (disk space not withstanding) strictly a benefit to the node. This idea is also somewhat orthogonal to CDN usage, as even if we did not opt to support CDNs a pack file format could still be used to speed up inter-node downloads.
+
+It is entirely possible to implement pack files as just a serialization of `HashMap<ObjectHash, Object>`.
+
+### Snapshots
+
+A feature currently on the roadmap becomes almost a non-issue with this approach; snapshotting or *checkpointing*, i.e. the ability to start a node without having to download all content from the network, requires no code modification in the sense that if objects are identified by immutable hashes, the local database can simply check if they are present to avoid redownloading them.
+
+Combined with pack files, a snapshot or *checkpoint* of a node is just a specific Block plus all of its dependencies in a single file.
+
+### Finality signatures
+
+[finality-signature]: #finality-signatures
+
+Finality signatures at first glance are an example of a "graph node" that points in the wrong direction: Each signature points to exactly one block, but they will not all be available at the time the block itself is created and its hash finalized. This would usually leave each signature "floating" around, not being a dependency of the block it points to:
+
+!["Floating" finality signatures](../diagrams/0008-finality-sigs-a.png)
+
+A possible solution to this problem is to define a window `k` during which finality signatures for a block can be attached to its child blocks. As an example, if block `n` is created, finality signatures for this block may be part of any block `(n+1)...(n+k)`.
```

Comment:
I don't understand why this is a problem in the first place. Why is it so important that everything is a dependency of a single ref?